library/core/src/panic/panic_info.rs

```diff
@@ -28,11 +30,29 @@ use crate::panic::Location;
 #[stable(feature = "panic_hooks", since = "1.10.0")]
 #[derive(Debug)]
 pub struct PanicInfo<'a> {
-    payload: &'a (dyn Any + Send),
+    payload: PayloadKind<'a>,
     message: Option<&'a fmt::Arguments<'a>>,
     location: &'a Location<'a>,
 }
 
+#[derive(Debug)]
+enum PayloadKind<'a> {
+    Any(&'a (dyn Any + Send)),
+    #[cfg(not(bootstrap))]
+    Error(&'a (dyn Error + 'static)),
+}
+
+impl<'a> PayloadKind<'a> {
+    fn downcast_ref<T: 'static>(&self) -> Option<&T> {
+        match self {
+            PayloadKind::Any(payload) => payload.downcast_ref(),
+            #[cfg(not(bootstrap))]
+            PayloadKind::Error(_) => None,
+            // PayloadKind::Error(error) => error.downcast_ref(),
```

Comment:
I kinda wanted to let this downcast for either but there isn't any reason to do so. I'd likely have to internally reimplement the `downcast_ref` logic to avoid needing to add `T: Error`.